tempesta_fw/http_parser.c

```diff
@@ -774,59 +771,47 @@ __FSM_STATE(RGen_HdrOtherV) {						\
  * those for parsing of message headers (__FSM_*), or for parsing
  * of message header values (__FSM_I_*, where _I_ means "interior",
  * or nested FSM). The major difference from __FSM_* macros is that
- * there can be zeroes in a message body that cannot be in a header.
+ * there is no need to fixup here.
  */
+#define __FSM_B_MOVE(to)			\
+	__FSM_MOVE_nff(to, 1, NULL, 0)
 #define __FSM_B_MOVE_n(to, n)						\
-do {									\
-	p += n;								\
-	if (unlikely(__data_offset(p) >= len)) {			\
-		/*							\
-		 * Postpone parsing until more data is available,	\
-		 * and start from state @to on next parser run.		\
-		 */							\
-		r = TFW_POSTPONE;					\
-		__fsm_const_state = to;					\
-		goto done;						\
-	}								\
-	c = *p;								\
-	goto to;							\
-} while (0)
-
-#define __FSM_B_MOVE(to)	__FSM_B_MOVE_n(to, 1)
+	__FSM_MOVE_nff(to, n, NULL, 0)
 
-#define TFW_HTTP_INIT_BODY_PARSING(msg, to_state)			\
-do {									\
+#define TFW_HTTP_PARSE_BODY()						\
+__FSM_STATE(RGen_Body) {						\
 	TFW_DBG3("parse msg body: flags=%#x content_length=%lu\n",	\
 		 msg->flags, msg->content_length);			\
 	/* RFC 2616 4.4: firstly check chunked transfer encoding. */	\
 	if (msg->flags & TFW_HTTP_CHUNKED)				\
-		__FSM_B_MOVE(to_state);					\
+		__FSM_B_MOVE(RGen_BodyChunk);				\
 	/* Next we check content length. */				\
 	if (msg->content_length						\
 	    && !(msg->flags & TFW_HTTP_VOID_BODY))			\
 	{								\
 		parser->to_read = msg->content_length;			\
-		__FSM_B_MOVE(to_state);					\
+		__FSM_B_MOVE(RGen_BodyChunk);				\
 	}								\
 	/* There is no body at all. */					\
+	msg->body.flags |= TFW_STR_COMPLETE;				\
 	r = TFW_PASS;							\
 	FSM_EXIT();							\
-} while (0)
-
-#define TFW_HTTP_PARSE_BODY()						\
-/* Read request|response body. */					\
-__FSM_STATE(RGen_Body) {						\
+}									\
+__FSM_STATE(RGen_BodyChunk) {						\
 	TFW_DBG3("read body: to_read=%d\n", parser->to_read);		\
+	__fsm_sz = __data_remain(p);					\
+	/* Open body for updading only if it were not opened */		\
+	if (!msg->body.ptr)						\
+		tfw_http_msg_add_data_ptr(msg, &msg->body, p, __fsm_sz);\
```

Comment:
`TFW_HTTP_UPDATE_BODY_IF_OPENED` check for `msg->body.ptr` not the `!msg->body->ptr`. This case is related to body opening.